org-code-javabuilder/lib/src/main/java/org/code/neighborhood/Painter.java

```diff
@@ -0,0 +1,146 @@
+        if(this.direction == null) {
+            throw new UnsupportedOperationException("Invalid direction given to painter");
+        }
+        this.remainingPaint = paint;
+        this.grid = World.getInstance().getGrid();
+        this.id = "painter-" + lastId++;
+    }
+
+    // Turn the painter one compass direction left (i.e. North -> West)
+    public void turnLeft() {
+        this.direction = this.direction.turnLeft();
+    }
+
+    // Move the painter one square forward in the direction the painter
+    // is facing
+    public void move() {
+        if (this.isValidMovement(this.direction)) {
+            if (this.direction.isNorth()) {
+                this.yLocation++;
+            } else if (this.direction.isSouth()) {
+                this.yLocation--;
+            } else if (this.direction.isEast()) {
+                this.xLocation++;
+            } else {
+                this.xLocation--;
+            }
+        } else {
+            System.out.println("You can't go that way");
+        }
+        System.out.println("New (x,y) : (" + this.xLocation + "," + this.yLocation + ")");
+    }
+
+    // Add paint of the given color to the grid at the location where the
+    // painter currently is
+    public void paint(String color) {
+        this.grid.getSquare(this.xLocation, this.yLocation).setColor(color);
+    }
+
+    // removes all paint on the square where the painter is standing
+    public void scrapePaint() {
+        this.grid.getSquare(this.xLocation, this.yLocation).removePaint();
+    }
+
+    // Returns how many units of paint are in the painter's personal
+    // bucket
+    public int getMyPaint() {
+        return this.remainingPaint;
+    }
+
+    // Hides the painter on the screen
+    public void hidePainter() {
+        System.out.println("You hid the painter");
+    }
+
+    // Shows the painter on the screen
+    public void showPainter() {
+        System.out.println("You displayed the painter");
+    }
+
+    // Painter adds a single unit of paint to their personal bucket
+    // Counter on the bucket on the screen goes down. If the painter
+    // is not standing on a paint bucket, nothing happens
+    public void takePaint() {
+        if (this.grid.getSquare(this.xLocation, this.yLocation).containsPaint()) {
+            this.grid.getSquare(this.xLocation, this.yLocation).collectPaint();
+            this.remainingPaint++;
+        } else {
+            System.out.println("There is no paint to collect here");
+        }
+    }
+
+    // Returns True if there is paint in the square where the painter
+    // is standing.
+    public boolean isOnPaint() {
+        return this.grid.getSquare(this.xLocation, this.yLocation).hasColor();
+    }
+
+    // Returns True if there is a paint bucket in the square where the
+    // painter is standing.
+    public boolean isOnBucket() {
+        return this.grid.getSquare(this.xLocation, this.yLocation).containsPaint();
```

Comment:
I think we might need a `hasBucket` method for a square--if there is a bucket with 0 paint in it, would we still want to return true here? My gut says yes, but I'm not sure. @jmkulwik

Reply:
How would we know if a square is a bucket?

Reply:
there's a bucket asset, although I'm not sure how we define that along with a paint count. This could be a follow-up item, might be good to discuss with curriculum what defines a bucket.

Reply:
I thought we decided that if there is no paint, the bucket disappears.

Reply:
That definitely simplifies things ^ 👍